server/src/main/java/org/candlepin/resource/OwnerProductResource.java

```diff
@@ -237,8 +237,13 @@ public ProductData getProduct(
     @Transactional
     public ProductCertificate getProductCertificate(
         @PathParam("owner_key") String ownerKey,
+        @ApiParam(name = "productId", required = true, value = "Numeric product identifier")
         @PathParam("product_id") String productId) {
 
+        if (!productId.matches("\\d+")) {
+            throw new BadRequestException(i18n.tr("Only numeric product IDs are allowed."));
```

Comment:
This feels like kind of a weird thing to expose/expect from callers. Would just returning null accomplish the same thing here without an error message that would probably garner a  "why?" from the caller?

Reply:
I would argue that receiving a null back from a request that is incorrect (since we can't generate product certificates from marketing products) would just delay the client's discovery of the error.

Reply:
In that case, would someone reading that error message connect the dots between "only numeric product IDs" and "marking products"?

Reply:
The reason it's a limitation is due to a poor design decision.  We embed the product ID into the OID used for an X509 extension.

```java
String productCertOid = OIDUtil.REDHAT_OID + "." +
    OIDUtil.TOPLEVEL_NAMESPACES.get(OIDUtil.PRODUCT_CERT_NAMESPACE_KEY);

// XXX need to deal with non hash style IDs
String productOid = productCertOid + "." + product.getId();
```

in `X509ExtensionUtil`.  OIDs can only consist of integers so deciding to encode the product ID into the OID saddled us with an otherwise unnecessary constraint.

I understand your point that the message I've written would probably just provoke a "Now what?" response from the user, but I'm not sure how else to phrase it.  The ideal would be to point them to some documentation, but we don't do that anywhere else in the code so there's no established convention for doing so.